spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/DatastoreQueryLookupStrategy.java

```diff
@@ -0,0 +1,108 @@
+import org.springframework.data.repository.query.QueryMethodEvaluationContextProvider;
+import org.springframework.data.repository.query.RepositoryQuery;
+import org.springframework.expression.spel.standard.SpelExpressionParser;
+import org.springframework.util.Assert;
+
+/**
+ * Query lookup strategy for Query Methods for Cloud Datastore.
+ *
+ * @author Chengyuan Zhao
+ *
+ * @since 1.1
+ */
+public class DatastoreQueryLookupStrategy implements QueryLookupStrategy {
+
+	private final DatastoreOperations datastoreOperations;
+
+	private final DatastoreMappingContext datastoreMappingContext;
+
+	private QueryMethodEvaluationContextProvider evaluationContextProvider;
+
+	private SpelExpressionParser expressionParser;
+
+	public DatastoreQueryLookupStrategy(DatastoreMappingContext datastoreMappingContext,
+			DatastoreOperations datastoreOperations,
+			QueryMethodEvaluationContextProvider evaluationContextProvider,
+			SpelExpressionParser expressionParser) {
+		Assert.notNull(datastoreMappingContext,
+				"A non-null DatastoreMappingContext is required.");
+		Assert.notNull(datastoreOperations,
+				"A non-null DatastoreOperations is required.");
+		Assert.notNull(evaluationContextProvider,
+				"A non-null EvaluationContextProvider is required.");
+		Assert.notNull(expressionParser, "A non-null SpelExpressionParser is required.");
+		this.datastoreMappingContext = datastoreMappingContext;
+		this.evaluationContextProvider = evaluationContextProvider;
+		this.datastoreOperations = datastoreOperations;
+		this.expressionParser = expressionParser;
+	}
+
+	@Override
+	public RepositoryQuery resolveQuery(Method method, RepositoryMetadata metadata,
+			ProjectionFactory factory, NamedQueries namedQueries) {
+		DatastoreQueryMethod queryMethod = createQueryMethod(method, metadata, factory);
+		Class<?> entityType = getEntityType(queryMethod);
+
+		if (queryMethod.hasAnnotatedQuery()) {
+			String sql = queryMethod.getQueryAnnotation().value();
+			return createGqlDatastoreQuery(entityType, queryMethod, sql);
+		}
+		else if (namedQueries.hasQuery(queryMethod.getNamedQueryName())) {
+			String sql = namedQueries.getQuery(queryMethod.getNamedQueryName());
+			return createGqlDatastoreQuery(entityType, queryMethod, sql);
+		}
+
+		throw new DatastoreDataException(
+				"Name-based query methods are not supported yet.");
+	}
+
+	@VisibleForTesting
+	<T> GqlDatastoreQuery<T> createGqlDatastoreQuery(Class<T> entityType,
+			QueryMethod queryMethod, String gql) {
+		return new GqlDatastoreQuery<>(entityType, queryMethod, this.datastoreOperations,
+				gql, this.evaluationContextProvider, this.expressionParser,
+				this.datastoreMappingContext);
+	}
+
+	@VisibleForTesting
+	Class<?> getEntityType(QueryMethod queryMethod) {
+		return queryMethod.getResultProcessor().getReturnedType().getDomainType();
+	}
+
+	@VisibleForTesting
+	DatastoreQueryMethod createQueryMethod(Method method, RepositoryMetadata metadata,
```

Comment:
What's the point of having this method? What does it add over just using `new DatastoreQueryMethod(method, metadata, factory)`?

Reply:
It's visible for testing as annotated. It's not practical to make a bunch of metadata that usually Spring Data generates from scanning a user's repo interface for a unit test.